storage/repair_queue.go

```diff
@@ -0,0 +1,133 @@
+func makeRepairQueue(storePool *StorePool, replicateQueue *replicateQueue, clock *hlc.Clock) repairQueue {
+	rq := repairQueue{
+		storePool:      storePool,
+		replicateQueue: replicateQueue,
+		clock:          clock,
+	}
+	// rq must be a pointer in order to setup the reference cycle.
+	rq.baseQueue = newBaseQueue("repair", &rq, repairQueueMaxSize)
+	return rq
+}
+
+// needsLeaderLease implements the queueImpl interface.
+// This should not require a leader lease, but if we find that there is too
+// much contention on the replica removals we can revisit that decision.
+func (rq repairQueue) needsLeaderLease() bool {
+	return false
+}
+
+// shouldQueue implements the queueImpl interface.
+func (rq repairQueue) shouldQueue(now proto.Timestamp, repl *Replica) (shouldQ bool, priority float64) {
+	shouldQ, priority, _ = rq.needsRepair(repl)
+	return
+}
+
+// needsRepair determines if any replicas that belong to the same range as repl
+// belong to a dead store. If so, it returns the priority (the number of
+// replicas on dead stores) and the list of these 'dead' replicas.
+func (rq repairQueue) needsRepair(repl *Replica) (bool, float64, []proto.Replica) {
+	rangeDesc := repl.Desc()
+	var deadReplicas []proto.Replica
+	for _, otherRepl := range rangeDesc.Replicas {
+		if rq.storePool.getStoreDetail(otherRepl.StoreID).dead {
+			deadReplicas = append(deadReplicas, otherRepl)
+		}
+	}
+
+	if len(deadReplicas) == 0 {
+		return false, 0, deadReplicas
+	}
+
+	// Does the range still have a quorum of live replicas? We shouldn't remove
+	// one without at least a quorum.
+	quorum := (len(rangeDesc.Replicas) / 2) + 1
+	liveReplicas := len(rangeDesc.Replicas) - len(deadReplicas)
+
+	// Priority is calculated based on need. By basing this off of the number
+	// of live replicas over he quorum amount ensures that in the case where a
+	// 5 replica range is only missing one it gets a lower priority than a 3
+	// replica range missing 1. Note that this may produce negative priorities,
+	// but these are correctly handled by baseQueue.
+	priority := float64(quorum - liveReplicas)
+	return liveReplicas >= quorum, priority, deadReplicas
+}
+
+// process implements the queueImpl interface.
+// If the replica still requires repairing, it removes the replicas on dead
+// stores (one at a time) and then queues up the replica for replication on the
+// repliateQueue.
+func (rq repairQueue) process(now proto.Timestamp, repl *Replica) error {
+	shouldQ, _, deadReplicas := rq.needsRepair(repl)
+	if !shouldQ {
+		// No more dead replicas, enqueue the replica in the replicate queue so
+		// any missing replicas can be added.
+		rq.replicateQueue.MaybeAdd(repl, rq.clock.Now())
+		return nil
+	}
+
+	// Only deal with one dead replica at a time.
+	if len(deadReplicas) < 1 {
+		return util.Errorf("no dead replicas, this shouldn't happen.")
+	}
+	deadReplica := deadReplicas[0]
+	log.Warningf("The replica %v is being removed from its dead store.", deadReplica)
```

Comment:
why %v? i'd be surprised if that produces a representation that's useful to a human